engine/engine.go

```diff
@@ -139,8 +139,9 @@ func (e *compatibilityEngine) NewRangeQuery(q storage.Queryable, opts *promql.Qu
 	}
 
 	lplan := logicalplan.New(expr, start, end)
-	if !e.disableOptimizers {
-		lplan = lplan.Optimize(logicalplan.DefaultOptimizers)
+	if e.logicalOptimizers == nil {
```

Comment:
Shouldn't this be the other way around i.e. `if e.logicalOptimizers != nil { apply them }`? Also, we need to take default optimizers into account. So, it should be something like the following if I am not mistaken:

```
var optimizers []logicalplan.Optimizers

if e.logicalOptimizers == nil {
  optimizers = append(optimizers, logicalplan.DefaultOptimizers)
}

optimizers = append(optimizers, e.logicalOptimizers...)
lplan = lplan.Optimize(optimizers)
``` 

And the same logic is missing in instant query function? Perhaps all of this logic could be moved into a separate function to avoid making a mistake?

Reply:
Good catch, thanks 👍 

I've added a method to determine which logical optimizers to use based on the input and inject them in the query. The query then just applies whatever it has in the logical optimizers variable.